scio-extra/src/main/scala/com/spotify/scio/extra/csv/CsvIO.scala

```diff
@@ -0,0 +1,231 @@
+  final val DEFAULT_WRITE_PARAMS = CsvIO.WriteParam(compression = beam.Compression.UNCOMPRESSED)
+
+  final case class ReadParam(
+    compression: beam.Compression = beam.Compression.AUTO,
+    csvConfiguration: CsvConfiguration = DEFAULT_CSV_CONFIG
+  )
+
+  final case class WriteParam(
+    compression: beam.Compression = beam.Compression.UNCOMPRESSED,
+    csvConfiguration: CsvConfiguration = DEFAULT_CSV_CONFIG,
+    suffix: String = ".csv",
+    numShards: Int = 1
+  ) {
+    def toReadParam: ReadParam =
+      ReadParam(compression = compression, csvConfiguration = csvConfiguration)
+  }
+
+  final case class Read[T: HeaderDecoder: Coder](path: String) extends ScioIO[T] {
+    override type ReadP = CsvIO.ReadParam
+    override type WriteP = CsvIO.WriteParam
+    final override val tapT = TapOf[T]
+
+    override protected def read(
+      sc: ScioContext,
+      params: ReadParam = CsvIO.DEFAULT_READ_PARAMS
+    ): SCollection[T] = CsvIO.read(sc, path, params)
+
+    override protected def write(data: SCollection[T], params: WriteP): Tap[T] =
+      throw new UnsupportedOperationException("Use CsvIO.Write() for writing")
+
+    override def tap(params: ReadP): Tap[T] = new CsvTap[T](path, params)
+  }
+
+  final case class Write[T: HeaderEncoder: Coder](path: String) extends ScioIO[T] {
+    override type ReadP = Nothing // WriteOnly
+    override type WriteP = CsvIO.WriteParam
+    final override val tapT = EmptyTapOf[T]
+
+    override protected def read(sc: ScioContext, params: ReadP): SCollection[T] =
+      throw new UnsupportedOperationException("Use CsvIO.Read() for reading")
+
+    override protected def write(data: SCollection[T], params: WriteP): Tap[Nothing] = {
+      data.applyInternal(csvOut(path, params))
+      EmptyTap
+    }
+
+    override def tap(params: ReadP): Tap[Nothing] = EmptyTap
+  }
+
+  final case class ReadWrite[T: HeaderCodec: Coder](path: String) extends ScioIO[T] {
+    override type ReadP = CsvIO.ReadParam
+    override type WriteP = CsvIO.WriteParam
+    final override val tapT = TapOf[T]
+
+    override protected def read(sc: ScioContext, params: ReadP): SCollection[T] =
+      CsvIO.read(sc, path, params)
+
+    override protected def write(data: SCollection[T], params: WriteP): Tap[T] = {
+      data.applyInternal(csvOut(path, params))
+      new CsvIO.CsvTap[T](path, params.toReadParam)
+    }
+
+    override def tap(params: ReadP): Tap[T] = new CsvIO.CsvTap[T](path, params)
+  }
+
+  private def csvOut[T: HeaderEncoder](path: String, params: WriteParam): FileIO.Write[Void, T] =
+    beam.FileIO
+      .write()
+      .to(path)
+      .withSuffix(params.suffix)
+      .withNumShards(params.numShards)
+      .withCompression(params.compression)
+      .via(new CsvSink(params.csvConfiguration))
+
+  private def read[T: HeaderDecoder: Coder](sc: ScioContext, path: String, params: ReadParam) =
+    sc.parallelize(Seq(path))
+      .readAll(beam.FileIO.matchAll())
```

Comment:
would it be possible to use a `.transform` here to composite all of these steps?